llms/mlx_lm/tuner/trainer.py

```diff
@@ -0,0 +1,204 @@
+    )
+    batch_size: int = field(default=4, metadata={"help": "Minibatch size."})
+    iters: int = field(default=100, metadata={"help": "Iterations to train for."})
+    val_batches: int = field(
+        default=25,
+        metadata={
+            "help": "Number of validation batches, -1 uses the entire validation set."
+        },
+    )
+    steps_per_report: int = field(
+        default=10,
+        metadata={"help": "Number of training steps between loss reporting."},
+    )
+    steps_per_eval: int = field(
+        default=200, metadata={"help": "Number of training steps between validations."}
+    )
+    steps_per_save: int = field(
+        default=100, metadata={"help": "Save the model every number steps"}
+    )
+    max_seq_length: int = field(
+        default=2048, metadata={"help": "Maximum sequence length."}
+    )
+    adapter_file: str = field(
+        default="adapter.npz",
+        metadata={"help": "Save/load path for the trained adapter weights."},
+    )
+
+
+def default_loss(model, inputs, targets, lengths):
+    logits, _ = model(inputs)
+    logits = logits.astype(mx.float32)
+
+    length_mask = mx.arange(inputs.shape[1])[None, :] < lengths[:, None]
+
+    ce = nn.losses.cross_entropy(logits, targets) * length_mask
+    ntoks = length_mask.sum()
+    ce = ce.sum() / ntoks
+
+    return ce, ntoks
+
+
+def iterate_batches(dset, tokenizer, batch_size, max_seq_length, train=False):
+    while True:
+        # Shuffle indices
+        indices = np.arange(len(dset))
+        indices = np.random.permutation(indices)
+
+        # Collect batches from dataset
+        for i in range(0, len(indices) - batch_size + 1, batch_size):
+            # Encode batch
+            batch = [tokenizer.encode(dset[indices[i + j]]) for j in range(batch_size)]
+            lengths = [len(x) for x in batch]
+
+            # Check if any sequence is longer than max_seq_length
+            if max(lengths) > max_seq_length:
+                print(
+                    "[WARNING] Some sequences are longer than 2048 tokens. "
+                    "Consider pre-splitting your data to save memory."
+                )
+
+            # Pad to the max length
+            batch_arr = np.zeros((batch_size, max(lengths)), np.int32)
+
+            for j in range(batch_size):
+                batch_arr[j, : lengths[j]] = batch[j]
+            batch = mx.array(batch_arr)
+            yield batch[:, :-1], batch[:, 1:], mx.array(lengths)
+
+        if not train:
+            break
+
+
+def evaluate(
+    model, dataset, loss, tokenizer, batch_size, num_batches, max_seq_length=2048
+):
+    all_losses = []
+    ntokens = 0
+    for it, batch in zip(
+        range(num_batches),
+        iterate_batches(dataset, tokenizer, batch_size, max_seq_length),
+    ):
+        losses, toks = loss(model, *batch)
+        all_losses.append((losses * toks).item())
+        ntokens += toks.item()
+
+    return np.sum(all_losses) / ntokens
+
+
+class LoraTrainer:
```

Comment:
This is a really general training loop, it might be better to name it `Trainer` rather than `LoraTrainer`.

I'm also wondering why you prefer it as a class? It doesn't seem like holding state is necessary and from an implementation and use standpoint it might be simpler to just have a `trainer.train` function (also more consistent with `evaluate`)

Reply:
Initially, my intention was to encapsulate all elements related to training within the trainer object. However, after making some changes, I feel that both a class and a function could be good options. Considering the `evaluate` used by testing, it now seems more logical to use the "train" function. I am happy to refactor it into a function.